app/packs/src/components/structureEditor/KetcherEditor.js

```
@@ -1,45 +1,845 @@
+        };
+      });
+    });
+    latestData.root.nodes = [...removeImagesFromData(latestData), ...imagesList_];
+  } catch (err) {
+    console.error("placeImageOnAtoms", err.message);
+  }
+};
+
+// generating images for ket2 format from molfile polymers list
+export const setKetcherData = async (rails_polymers_list, data) => {
+  let collected_images = [];
+  if (rails_polymers_list && rails_polymers_list.length) {
+    const { c_images, molfileData, image_counter } = await adding_polymers_ketcher_format(rails_polymers_list, mols, data, image_used_counter);
+    image_used_counter = image_counter;
+    molfileData?.root?.nodes.push(...c_images);
+    return { collected_images: c_images, molfileData };
+  }
+  return { collected_images, molfileData: data };
+};
+
+// helper function to test alias list consistency 0,1,2,3,4...
+export const isAliasConsistent = () => {
+  const index_list = [];
+  const uniqueIndices = new Set();
+
+  for (const mol of mols) {
+    const molecule = latestData[mol];
+    molecule?.atoms?.forEach((item) => {
+      if (item.alias) {
+        const splits = item.alias.split("_");
+        const index = parseInt(splits[2]);
+
+        // Check for duplicates
+        if (uniqueIndices.has(index)) {
+          return false;  // Duplicate found
+        }
+        uniqueIndices.add(index);
+        index_list.push(index);
+      }
+    });
+  }
+
+  index_list.sort((a, b) => a - b);
+  for (let i = 0; i < index_list.length; i++) {
+    if (index_list[i] !== i) {
+      return false;  // Missing or incorrect number sequence
+    }
+  }
+  return true; // Passed all checks
+};
+
+// IMP: helper function to handle new atoms added to the canvas
+export const handleAddAtom = async () => {
+  let already_processed = [];
+  image_used_counter = -1;
+  const seenThirdParts = new Set();
+
+  for (let m = 0; m < mols.length; m++) {
+    const mol = latestData[mols[m]];
+    for (let a = 0; a < mol?.atoms?.length; a++) {
```

- If mol or mol.atoms is null or undefined, this loop will silently do nothing, potentially masking issues.
- Suggestion: Validate the structure of mol and mol.atoms before entering the loop.